### ISSUE TITLE
Rancher-monitoring installation fails on fresh cluster with crd error

### ISSUE THREAD

**Issue by bashofmann** [rancher/rancher#32827]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
* Create a new Rancher cluster
* Install rancher-monitoring into the local cluster

This was reproduced on a fresh rke2 cluster.

or

* Create a new downstream cluster
* Install rancher-monitoring into the downstream cluster

**Result:**
The job that creates the CRDs fails with:

```
rancher-monitoring-crd-create-w4bnw create-crds Resource: "apiextensions.k8s.io/v1, Resource=customresourcedefinitions", GroupVersionKind: "apiextensions.k8s.io/v1, Kind=CustomResourceDefinition"
rancher-monitoring-crd-create-w4bnw create-crds Name: "prometheuses.monitoring.coreos.com", Namespace: ""
rancher-monitoring-crd-create-w4bnw create-crds for: "/etc/config/crd-manifest.yaml": CustomResourceDefinition.apiextensions.k8s.io "prometheuses.monitoring.coreos.com" is invalid: spec.preserveUnknownFields: Invalid value: true: must be false in order to use defaults in the schema

rancher-monitoring-crd-create-w4bnw create-crds Resource: "apiextensions.k8s.io/v1, Resource=customresourcedefinitions", GroupVersionKind: "apiextensions.k8s.io/v1, Kind=CustomResourceDefinition"
rancher-monitoring-crd-create-w4bnw create-crds Name: "alertmanagers.monitoring.coreos.com", Namespace: ""
rancher-monitoring-crd-create-w4bnw create-crds for: "/etc/config/crd-manifest.yaml": CustomResourceDefinition.apiextensions.k8s.io "alertmanagers.monitoring.coreos.com" is invalid: spec.preserveUnknownFields: Invalid value: true: must be false in order to use defaults in the schema
```

**Other details that may be helpful:**
Monitoring was deployed directly after the cluster was created, e.g. with terraform. So, it could be some kind of race condition.
The issue started to appear first with Rancher 2.5.8.

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): 2.5.8
- Monitoring version:  14.5.100
- Installation option (single install/HA): HA

**Comment by aiyengar2** [comment]
An explanation for why this could be appearing sporadically in some clusters but not consistently reproducible in all setups could be because we do not currently wait for CRDs to be established before continuing to the next steps in our CRD chart.

This would explain this specific error since a fresh cluster with Rancher Monitoring V1 CRDs (which is based on apiextensions v1/beta1) requires the `set-preserve-unknown-fields-false` initContainer to successfully update the CRD before the apiextensiosn v1 CRD is applied.

We should re-test this issue to see if we can reproduce it after the fix in https://github.com/rancher/charts/pull/1203 is merged.

**Comment by bashofmann** [comment]
@aiyengar2 Now that rancher/charts#1203 has been merged, I tested if this solves the problem. I can still reliably reproduce this issue on

* 2.5.8 with 14.5.100
* v2.5-head with 14.5.101+up14.5.0 (out of the dev-v2.5 branch of the charts repo)

Or did I test too early and the relevant change is not included yet?

**Comment by bashofmann** [comment]
I tried it with master-head now, the installation runs through. But the `is invalid: spec.preserveUnknownFields: Invalid value: true: must be false in order to use defaults in the schema` error is still present in the logs of the job that creates the crds.

[master-head.log](https://github.com/rancher/rancher/files/6625305/master-head.log)

**Comment by aiyengar2** [comment]
Hmm you are right @bashofmann, I was able to reproduce this. Moving this back to Working

**Comment by PPCM** [comment]
Do you still plan to create a backport issue for 2.5.9?

**Comment by aiyengar2** [comment]
@PPCM the backport issue is in https://github.com/rancher/rancher/issues/33014.

**Comment by serverbaboon** [comment]
No sure if this is useful information but I can produce the above when using Fleet to install the Rancher Apps fairly consistently.

I have a Fleet repo that will install Monitoring, Logging, CertManager, Autoscaler (not Rancher) and CIS scanner and everything works ok as the installs sort themselves out.

If I install istio (say 1.9.6) by adding a label to trigger another repo everything still ok.

However if apply the istio label before assigning the cluster to a workspace so that Istio is deployed at the same time there seems to be a clash (come on bundle dependancy!!) which cause a failure in a rancher-monitoring-crd, missed whether its the Monitoring deployment CRD or the Istio monitoring crd deployment.

**Comment by aiyengar2** [comment]
@serverbaboon the root cause of this that was assessed in https://github.com/rancher/charts/pull/1203 and https://github.com/rancher/charts/pull/1278. There were two issues:
1. Our CRD install job was not waiting to see whether the CRDs were installed **and** established in the API server before continuing installation. It was only firing an installation and moving forward. Therefore, in situations where you were automating the install of the chart (Terraform, Fleet), where rancher-monitoring-crd was quickly followed by an installation of rancher-monitoring, you were more likely to see errors on rendering the template since the updated CRD was not established yet.
2. The CRD install initContainer that allows you to update CRDs within a cluster that were based on non-structural schemas (e.g. `preserveUnknownFields=true`) was silently failing to actually update the CRD within the cluster, as shown in the logs of https://github.com/rancher/charts/pull/1278.

Based on your description, it seems like the conflict you are experiencing is possibly due to the prior, right? Or are you referring to a new issue?

**Comment by serverbaboon** [comment]
Sorry I thought I had replied.

Yes I am guessing it is, I just seemed to see it more when Istio was part of the same installs rather than monitoring on its own.

**Comment by snasovich** [comment]
@cbron , is this something Brenda or Jack can take care of validating since it's `status/dev-validate`?

**Comment by black444** [comment]
Do you have any workaround? I've installed Rancher 2.5.9, but issue still exists.

**Comment by black444** [comment]
I waited for release 2.5.9 which was supposed to fix this, but problem still occurs. Anyway, found workaround which is working fine for me. First, I installing monitoring with default values which allows rancher-monitoring-crd to be installed successfully. Then I'm uninstalling it and installing it again, but this time with my custom parametrization (pvc, resources).

**Comment by snasovich** [comment]
@black444 , unfortunately, this fix didn't make it to 2.5.9 as the backport issue (#33014) targeted 2.5.10 only.
Thank you for suggesting the workaround.

**Comment by snasovich** [comment]
@brendarearden or @jiaqiluo , would be great if one of you two can validate this to help QA? Thank you.

**Comment by jiaqiluo** [comment]
the fix is validated in rancher v2.6-head e3316c9 with Monitoring v2 chart `100.0.0+up16.6.0`

Steps:
- run rancher in single-install mode 
- provision multiple RKE/k3s clusters with various k8s versions and number of nodes
- install monitoring v2 into the cluster via the cluster explorer UI

Results:
- monitoring is installed successfully without hitting the error mentioned in the issue
